### PR TITLE
Thread.yield fairness with multiple busy yielding threads

### PR DESCRIPTION
**Pull request by ahh**
Here's a very hacky repro for the problem: the below code counts how many times a thread goes through a loop yielding before successfully *doing* so.  At trunk, this runs 10,000+ times per thread, or more if the threads are pinned to the same CPU.  With this patch, we reliably yield every time, because a yielding thread is properly marked as a waiter.
```
let threads = 4
let print = Sys.argv.(1) |> bool_of_string
let iters = Sys.argv.(2) |> int_of_string

let yields = ref 0
let are_ready = ref 0
let are_done = ref 0

let last = ref '!'

let start = ref 0.0

let stop = ref 0.0

let run_lengths = ref []

let report label cnt =
  let old = !run_lengths in
  run_lengths := (label, cnt) :: old;;

let threads =
  List.init threads (Thread.create (fun i ->
    (* Label our threads A..Z (clearer than TIDs, and integer indices are confusable
       with run lengths) *)
    let label = Char.unsafe_chr (Char.code 'A' + i) in
    incr are_ready;
    (* Don't make any progress until all threads are spawned and properly contending for
       the Ocaml lock. *)
    if !are_ready = threads then start := (Unix.gettimeofday ());
    while !are_ready < threads do
      Thread.yield ()
    done;
    let consecutive = ref 0 in
    while !yields < iters do
      incr yields;
      last := label;
      Thread.yield ();
      incr consecutive;
      if not (!last = label)
      then (
        report label !consecutive;
        consecutive := 0)
    done;
    if !consecutive > 0 then report label !consecutive;
    incr are_done;
    if !are_done = threads then stop := (Unix.gettimeofday ())
  ));;


List.iter Thread.join threads;;

let report (label, count) = Printf.printf "%c ran for %d\n" label count;;

if print then List.iter report (List.rev !run_lengths)

let time = !stop -. !start;;
let switches = List.length !run_lengths;;
Printf.printf "Executed in %fs with %d switches\n" time switches

```

I can measure no overhead in throughput from the normal volume of yielding due to the tick thread (say, with two threads merrily allocating) but I'm happy to run more tests.

Thanks!

### CODE REVIEW

**Comment by xavierleroy** [comment]
Interesting proposal, thanks!  @jhjourdan, would you like to look at it?

**Comment by xavierleroy** [comment]
For easy reference: [MPR 7669](https://caml.inria.fr/mantis/view.php?id=7669) and GPR #1533.

**Comment by jhjourdan** [comment]
Interesting proposal, indeed! I agree that this approach is much more reliable than my dirty nanosleep hack, hence I am not surprised by your fairness measurements.

@xavierleroy, do you think we could add the code snippet above to the test suite? Or do you think these kind of measurements are not reliable enough?

Moreover, it seems the Windows API now supports condition variables. We could re-implement the master lock using those, and hence have fairness on Windows too. They are not supported under Windows Server 2003 and Windows XP, so that we would need to fallback to the current implementation on XP or even drop support for XP (do people still care about Windows XP nowdays ?).

Apart from that (and what is shown by the continuous integration), I reviewed the code, and it seems OK to me.

**Comment by dra27** [comment]
If it’s one or two little `#ifdef`s to support Windows XP/2003 then fine to keep the support, but otherwise don’t worry - building OCaml on XP/2003 is already very difficult to set up, so the person/company intent on doing so will probably also be able to restore the old code!

**Comment by dra27** [comment]
(There is a reasonable chance that multicore will absolutely force Windows 7+ anyway, which will be most welcome...)

**Comment by ahh** [comment]
@jhjourdan I hadn't added an explicit test because yield behavior is so nondeterministic, but now that I think more it's actually not so bad.  I added a slightly simplified version of that test above, take a look.  I don't really know how the test metadata works and I can't find a reference to it, so I'm not sure the comment block at the top of the test is correct, but it seems to pass as is and fail if I break it, so that's good.

(I have no objection to dropping the test, but if you want it, here you go.)

**Comment by ahh** [comment]
(Apologies, fatfingered "close".)

**Comment by jhjourdan** [comment]
> @jhjourdan I hadn't added an explicit test because yield behavior is so nondeterministic, but now that I think more it's actually not so bad. I added a slightly simplified version of that test above, take a look. I don't really know how the test metadata works and I can't find a reference to it, so I'm not sure the comment block at the top of the test is correct, but it seems to pass as is and fail if I break it, so that's good.

Great. I am personally in favor of adding this test. We need to know what others think. To make sure that we won't get any false negatives, we could have a better safety margin (i.e., use a threshold of 10 instead of 3). 

BTW, there are still complains of the CI about too long lines.

**Comment by jhjourdan** [comment]
@dra27 : any idea why the Appveyor CI is failing ? This seems completely unrelated to this PR.

**Comment by nojb** [comment]
From the log:
```
[00:07:11] i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
[00:07:11]    -o st_stubs_b.o st_stubs.c
[00:07:11] In file included from st_stubs.c:52:0:
[00:07:11] st_win32.h:104:36: error: unknown type name ‘st_masterlock’
[00:07:11]  static INLINE void st_thread_yield(st_masterlock * m)
[00:07:11]                                     ^~~~~~~~~~~~~
[00:07:11] st_stubs.c: In function ‘caml_thread_yield’:
[00:07:11] st_stubs.c:748:3: error: implicit declaration of function ‘st_thread_yield’ [-Werror=implicit-function-declaration]
[00:07:11]    st_thread_yield(&caml_master_lock);
[00:07:11]    ^~~~~~~~~~~~~~~
[00:07:11] cc1: all warnings being treated as errors
[00:07:11] make[3]: *** [Makefile:96: st_stubs_b.o] Error 1
```

**Comment by jhjourdan** [comment]
> If it’s one or two little #ifdefs to support Windows XP/2003 then fine to keep the support, but otherwise don’t worry - building OCaml on XP/2003 is already very difficult to set up, so the person/company intent on doing so will probably also be able to restore the old code!

I guess that's something to discuss with the dev team. For now, I will prepare a PR for using condition variables on > XP using ifdefs, once this one will get merged.

**Comment by ahh** [comment]
Fixed the long lines. Think I fixed the AppVeyor build but less certain of that.

**Comment by ahh** [comment]
I uploaded a fix for the Windows build, but the pull request isn't updating; I'm not sure why. https://github.com/ahh/ocaml/commit/c8508627d1c6c42a3b678ae6c00729eedd5a95ea shows the log

**Comment by jhjourdan** [comment]
The Windows build shows the following error message:

```
i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
   -o st_stubs_b.o st_stubs.c
In file included from st_stubs.c:52:0:
st_win32.h: In function ‘st_thread_yield’:
st_win32.h:161:5: error: expected ‘;’ before ‘}’ token
     }
     ^
```

**Comment by jhjourdan** [comment]
The Windows build shows the following error message:

```
i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
   -o st_stubs_b.o st_stubs.c
In file included from st_stubs.c:52:0:
st_win32.h: In function ‘st_thread_yield’:
st_win32.h:161:5: error: expected ‘;’ before ‘}’ token
     }
     ^
```

**Comment by jhjourdan** [comment]
The Windows build shows the following error message:

```
i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
   -o st_stubs_b.o st_stubs.c
In file included from st_stubs.c:52:0:
st_win32.h: In function ‘st_thread_yield’:
st_win32.h:161:5: error: expected ‘;’ before ‘}’ token
     }
     ^
```

**Comment by jhjourdan** [comment]
The Windows build shows the following error message:

```
i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
   -o st_stubs_b.o st_stubs.c
In file included from st_stubs.c:52:0:
st_win32.h: In function ‘st_thread_yield’:
st_win32.h:161:5: error: expected ‘;’ before ‘}’ token
     }
     ^
```

**Comment by jhjourdan** [comment]
The Windows build shows the following error message:

```
i686-w64-mingw32-gcc -c -O -mms-bitfields -Wall -Wno-unused -fno-tree-vrp -Werror -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../../runtime  \
   -o st_stubs_b.o st_stubs.c
In file included from st_stubs.c:52:0:
st_win32.h: In function ‘st_thread_yield’:
st_win32.h:161:5: error: expected ‘;’ before ‘}’ token
     }
     ^
```

**Comment by dra27** [comment]
CIs both kicked...

**Comment by dra27** [comment]
Ok, the failure now looks to be something to do with the new test - I’ll try to have a look later, unless someone else has access to a Windows box in the meantime?

**Comment by jhjourdan** [comment]
The test should be disabled on Windows. There is no fix for this issue on Windows (yet ?), so there is no reason for this test to pass.

**Comment by damiendoligez** [comment]
@dra27 I'm looking into it.

**Comment by jhjourdan** [comment]
@damiendoligez : Again, I don't think there is anything to do here: `st_win32.h` does not contain the fix proposed by this PR, because it needs condition variables, which are not available on Windows XP.

At least for now, I think we should just disable the test on Windows, just like we did for #1533.

**Comment by ahh** [comment]
I agree that it's probably best to disable the test on Windows, but it's at least a *little* surprising it doesn't work; note that Windows sleeps unconditionally since we don't have a waiter count there.  That still leaves a lot of ways for the scheduler to hose us (deciding to wake up the waiter on a remote CPU which takes too long to run, for example!) though.

**Comment by damiendoligez** [comment]
Looks like the Windows scheduler is not as reliable as the Linux one. I'm getting some test runs that are OK, but most of the time I get a handful of random numbers. The highest I've observed so far is 5598.

Looks like disabling the test under Windows is the best solution.

**Comment by xavierleroy** [comment]
Regarding Windows: I was under the impression that the Windows scheduler is rather fair, and that the `Sleep(0);` seems to yield to other threads like we expect it to do.  So, I'd rather stick with the current Win32 code rather than embark on a reimplementation with condition variables and the like.  

Now, the hiccups that @damiendoligez reports maybe prove me wrong, or maybe they are exacerbated by our test setting (virtual machines running on a heavily-loaded Jenkins/Cloudstack installation).

I still think we should not include the test in our test suite, even under Linux, because it is testing something (a latency) that is not reproducible (especially on heavily-loaded hardware).  A test should pass all the time or fail all the time.

**Comment by jhjourdan** [comment]
> Now, the hiccups that @damiendoligez reports maybe prove me wrong, or maybe they are exacerbated by our test setting (virtual machines running on a heavily-loaded Jenkins/Cloudstack installation).

We already observed that when writing the test for #1533. We disabled it on Windows because `Sleep(0)` was not able to yield reliably to another thread.

> I still think we should not include the test in our test suite, even under Linux, because it is testing something (a latency) that is not reproducible (especially on heavily-loaded hardware). A test should pass all the time or fail all the time.

We are *not* testing a latency here. The result of this test should not depend on whether the hardware is heavily-loaded or not. Using condition variables, each call to `yield` *will* give the control to another thread. The only case that would make the test fail is when the thread that receives the control gets immediately interrupted by the scheduler. Moreover, if we use a good safety margin in the number of observed yield failure, this has to happen several times in a row, which is highly unlikely (if not impossible, since there is only a limited number of code points in the loop that would allow handling the scheduling signal).

**Comment by ahh** [comment]
I don't :(

I was expecting the new test to pass on Windows because windows
unconditionally sleeps, but I don't have any idea at all about Window's
scheduling behavior. We could just disable the test on windows until we
merge a condition-variable based lock there.

**Comment by ahh** [comment]
Uploaded a patch that disables the test on Windows (if I understand the metadata language correctly.)

**Comment by jhjourdan** [comment]
Great, CI succeeds!

Seems like the only thing left to decide about this PR is whether or not to include the test. I don't have a strong opinion about that, even though I think that having it would be better. Fairness is quite fragile, and not having a test for it would prevent us to see if some other update breaks it in the future. That said, I do understand @xavierleroy's concerns about robustness of the test suite.

**Comment by ahh** [comment]
I don't have a strong opinion either, but I also favor leaving it in.  I would be quite surprised if it proved flaky as is, and it's not hard to remove it if we discover it does (for non-fixable reasons.)

(Assuming the implementation of yield behaves as we believe it does, for worker thread A running the test to run two consecutive iterations requires either:

(1) a non-worker thread to execute
(2) another worker thread to drop the ocaml lock somewhere other than a yield
(3) a spurious condvar wakeup

(1) shouldn't happen other than maybe main thread once while setting up join.  (2) seems unlikely since the loop only calls yield and modifies int refs (it certainly doesn't allocate!). As for (3), while Posix allows them, I suspect most modern implementations will never cause one in current usage--I think they are allowed mostly for reasons involving the memory model.  They're certainly very rare.

For the record, on my system the test seems to consistently pass with a limit of 1.  That seems a bit conservative, but I do not expect any flakiness with any reasonable threshold.)

So I'd suggest we merge the test, but keep an eye out; if it proves flaky, disable or remove it (and investigate why!)

**Comment by ahh** [comment]
Someone with a commit bit want to make a decision? :)

**Review by gasche — APPROVED** [review]
Approved on the basis of @jhjourdan's review.

I would be in favor of keeping this test in, and disabling the first time we observe it breaking when it should not.

**Comment by damiendoligez** [comment]
I'm also in favor of keeping the test.

**Comment by ahh** [comment]
I updated the PR to the head of trunk. I think this is ready to go?  Being new here I am not exactly sure what the next step is; do we need to do anything before merging this?

**Comment by xavierleroy** [comment]
I would have liked a second technical review, and I don't have time to do it myself right now. 

I'd like to be absolutely sure there is no possibility of deadlock (a thread suspending itself on the "wake up" condition variable and never waked up). 

Also, there is a shared variable that is accessed both within a locked section and outside, not unlike the infamous "double checked locking" Java anti-pattern.  I'd like confirmation that the race is harmless even with a very relaxed memory model.

**Comment by ahh** [comment]
>  I'd like confirmation that the race is harmless even with a very relaxed memory model.

A few comments on the double-checked logging:

(1) We were already doing this (implicitly) before this PR--I didn't change where we looked at the value of `m->waiters`, and the new yield function modifies it under the same lock.  So the situation is not worse than it was before, in any case.

(2) As a sometimes-member of the C++ committee concurrency group, I will in fact say that the program as written is *not* safe: we invoke a data race when we read m->waiters that might be modified concurrently by another thread (and volatile does not save you here!)  In the C/C++ memory model, this is undefined behavior.  The correct solution, in C11-compliant code, would be to invoke `atomic_load` and `atomic_store` with `memory_order_relaxed` on an `atomic_int`.  But that is orthogonal to the change being made here--see point 1--and is mostly a long term issue for correctness with more modern compilers (and analysis tools.) In practice you'll get away with it here (as much, as a concurrency person, I hate to admit it.)

(3) Double-checked locking is a badly overloaded term. The Java anti-pattern you refer to specifically considers an (unlocked) read indicating that other (lock-protected) data is ready to consume.  This is why the memory model matters in that case: a thread initializing data like so

```
x = 1
y = 1
z = &x
```
appears to guarantee, to a naive reader, that if a consuming thread sees `y == 1` or `z == &x` that certainly the same thread sees `x == 1`, but this just isn't so without memory barriers (or architectural ordering guarantees.)  But that's not what we're doing here: the read of `waiters` outside a lock isn't used to gate accessing any other data outside a lock, so there can't be any ordering issues.  (An easy way to see this: depending on the value of `waiters` we either do A) nothing at all or B) take the lock and do things--so we can't possibly be racily accessing other data.) The code here is--or should be, at least :) -- correct in an arbitrarily weak memory model. 

In short: the double checked locking should not be an issue.

**Comment by ahh** [comment]
As for potential deadlock, I believe this informal proof holds. (The analysis is slightly clearer, and doesn't change the result, if you imagine threads wrote their own TID instead of 1 into `m->busy`, so that we can immediately read the "current" thread off of the state of the masterlock. As a side note, this might be a good idea to actually *do* for debugging purposes, and it'd be quite cheap.)

(0) We never call `st_masterlock_acquire` twice in the same thread without a corresponding `st_masterlock_release`. (That'd be obviously broken and deadlock in any version of the code.)

(1) We don't block on the condvar in `st_masterlock_acquire` unless t->busy tells us a *different* thread holds the lock.  (i.e. we can assume that if we block on the condvar there, then `t-busy` is nonzero *and* not equal to our tid.)

(2) We don't block on the condvar in `st_thread_yield` without setting `t->busy = 0` first.

(3) Therefore, if `m->busy == t`, then `t` is not blocked on the condvar.  (The only two places we block are covered by (1) and (2).)

(4) At the beginning of any critical section of `m->lock`, if we observe `m->waiters > 0` then at least one thread is waiting on the condvar.  (Obvious by inspection: we only increase `m->waiters` directly before waiting on the condvar.)

(5) To reach a deadlock, we'd have to have every thread simultaneously in a call to `pthread_cond_wait`. (Threads in a caml_blocking_section off in a syscall somewhere aren't really important here--ignoring them for simplicity.)

(6) Consider any such call to `pthread_cond_wait`. It was either from `yield` or `acquire`.
(a) If it was from `yield`, it first observed `m->waiters > 0` and called `pthread_cond_signal`. By (4) at least one thread was waiting on the condvar, which certainly will be woken by our signal.
(b) If it was from `acquire`, it first observed `m->busy = t` for some nonzero t (not itself.)  By (3) that thread isn't blocked.

In either case at least one thread is always awake (again, not counting the case that no one is attempting to run and all threads are blocked inside a blocking section's syscall, but that's not relevant.)  So we can't deadlock.

Seem reasonable?

**Comment by jhjourdan** [comment]
> So the situation is not worse than it was before, in any case.
> I will in fact say that the program as written is not safe:
> In short: the double checked locking should not be an issue.

Still, I think @xavierleroy has a point: This program is not safe under the C11 memory model. Not that I think this should be avoided at all costs (the OCaml runtime is written in C99 where we do not have atomic variables), but that should be avoided when we can.

And, as it turns out, I suspect the first check of the waiters count is not really useful here. If there is only one active thread, then taking and releasing the lock should essentially be free, since the mutex's data structures should belong to the core's cache. Am I right in my reasoning?

**Comment by jhjourdan** [comment]
(That would actually mean we could get rid of the `st_masterlock_waiters`, which is good since it is broken on Windows.)

**Comment by jhjourdan** [comment]
> Seem reasonable?

It seems to me, at least.

**Comment by ahh** [comment]
> Still, I think @xavierleroy has a point: This program is not safe under the C11 memory model. Not that I think this should be avoided at all costs (the OCaml runtime is written in C99 where we do not have atomic variables), but that should be avoided when we can.

I agree, but I want to emphasize that I didn't make this unsafe.  Any place where what we're doing is unsafe and might cause problems would also be unsafe right now.  

> And, as it turns out, I suspect the first check of the waiters count is not really useful here. If there is only one active thread, then taking and releasing the lock should essentially be free, since the mutex's data structures should belong to the core's cache. Am I right in my reasoning?

Unfortunately you are not.  Even an uncontended mutex acquire can be surprisingly noticeable; atomic instructions aren't free. (They are much cheaper than they were, say, fifteen years ago, but that's not the same as free.)  I would be leery of removing the waiters check.

**Comment by jhjourdan** [comment]
> Unfortunately you are not. Even an uncontended mutex acquire can be surprisingly noticeable; atomic instructions aren't free.

I've just done some tests on my machine:

```C
#include <pthread.h>
#include <unistd.h>
#include <stdio.h>

int ended = 0;
void* thread1(void* arg) {
  sleep(10);
  ended = 1;
}

pthread_mutex_t lock;
int n2 = 0;

void* thread2(void* arg) {
  while(!ended) {
    pthread_mutex_lock(&lock);
    pthread_mutex_unlock(&lock);
    n2++;
  }
}

int main (void) {
  pthread_t t;
  pthread_create(&t, NULL, thread1, NULL);

  pthread_mutex_init(&lock, NULL);

  // pthread_create(&t, NULL, thread2, NULL);

  int n1 = 0;
  while(!ended) {
    pthread_mutex_lock(&lock);
    pthread_mutex_unlock(&lock);
    n1++;
  }

  printf("%d %d\n", n1, n2);

  return 0;
}
```

It seems like that if only one thread is using the lock, then a full acquire-release cycle takes around 50 cycles. Do you really think 50 cycles are too much for Thread.yield to afford? Or do you think this could be more on other architectures?

**Comment by ahh** [comment]
I do think it could be more on other architectures; it's hard to speak authoritatively without hardware to benchmark at hand, but in the past architects I've spoken to have generally not let me pin them down on the cost of an uncontended atomic operation, because it's often hard for them to _make_ those guarantees. :)  But that said, honestly, I do also think 50 cycles is a bad thing to accept here.

In particular: in Async, we yield at certain points that are fairly hot, because doing so can dramatically improve fairness (and, though the experiments here predate me, I believe also reduce the effective latency of long-running syscalls.)  When we need to yield, doing so here is very effective; when we don't need to, yielding is *free*. So it's a very easy decision.

Adding any significant cost--even just a pthread lock!--makes it an actual, possibly difficult decision. Now we don't know what's worth it and what's not; we'll have to make judgment calls we could get wrong; the situation isn't an automatically one-sided tradeoff. Would it still be worth doing in Async? Probably, but that's not the point. Right now we can unconditionally just yield and not worry about it in the case we won't yield.  That seems very valuable to me.  

And to re-iterate, the memory model issues aren't changed by this patch set at all.  We do exactly the same locking and accesses.  If we want to have a separate conversation about fixing that, we can, but I don't think it's necessary to do it here.

Anyway, that's my two cents.

**Comment by xavierleroy** [comment]
Concerning the unprotected read from `m->waiters`:

I know there's a race condition there in the original code; I might even have written it myself.  The reasoning at the time was that if the value read from `m->waiters` is sometimes wrong (incl. out of thin air) we'll skip some yields or do some useless yields, but nothing really wrong will happen.  

I think the new code has the same property, because of the lock-then-check-again trick, but I wanted to make sure this is the case.

As regards C11, yes, the race is undefined behavior, but I still hope that GCC and Clang refrain from producing crazy code here.

**Comment by xavierleroy** [comment]
Concerning the optimization "if no waiters then don't even release and acquire the master lock":

I'm pretty sure this optimization was not there initially and was added later based on observations by serious users that it makes a noticeable difference in practice.  I'd need to dig in the bug tracker and my old e-mail to find out the OS and the kind of application where the difference was observed.

**Comment by ahh** [comment]
> I know there's a race condition there in the original code; I might even have written it myself. The reasoning at the time was that if the value read from m->waiters is sometimes wrong (incl. out of thin air) we'll skip some yields or do some useless yields, but nothing really wrong will happen.

With the exception of memory model issues, I agree with this reasoning entirely.

> I think the new code has the same property, because of the lock-then-check-again trick, but I wanted to make sure this is the case.

I agree that reading a zany value from m->waiters outside the lock is still fine--we maybe miss a yield, or maybe take a lock unnecessarily, but that's fine.

(The second check under the lock *is* necessary to prevent deadlock, obviously.)


> As regards C11, yes, the race is undefined behavior, but I still hope that GCC and Clang refrain from producing crazy code here.

I think, in practice, you'll be mostly OK. I *do* think that moving to C11 and doing it right is worth doing when the portability issues are satisfactory, but it's not a high priority.

**Comment by damiendoligez** [comment]
@ahh you should add @jhjourdan's name in the list of reviewers, then I think it's good to merge.

**Comment by ahh** [comment]
@damiendoligez Can I do that as a non-maintainer? Sorry, I don't see a button.

**Comment by dbuenzli** [comment]
@ahh he meant in the changelog credits [here](https://github.com/ocaml/ocaml/pull/2112/files#diff-c112bb3542e98308d12d5ecb10a67abcR509).

**Comment by ahh** [comment]
Ah, of course, that makes more sense. :)  Done!

**Comment by ahh** [comment]
Okay, sorry, I really badly screwed up that rebase and fix, and then screwed it up again trying to fix it. I think all is fine now--we have exactly the test from  88bde14, a Changes that mentions @jhjourdan, and the same code we've been happy with.  PTAL.  Sorry for the line noise here.

**Comment by damiendoligez** [comment]
Thanks !

**Comment by ahh** [comment]
Thanks!

**Pull request merged**
